### PR TITLE
[prereq] Pull out Balanced logic to a BalancedScoreTracker

### PR DESCRIPTION
**Pull request by iamdanfox**
## Before this PR

I have a big old PR here https://github.com/palantir/dialogue/pull/853, so just want to land a refactor independently to make the diff easier to read.

## After this PR
==COMMIT_MSG==
Pull out Balanced logic to a BalancedScoreTracker class
==COMMIT_MSG==

## Possible downsides?

- this does chop out the BALANCED_RTT2 functionality, but I think this is OK because we had to turn it off everywhere and it's gonna need a little more work to get it into a useable state. figured there's not much point polishing somethng that isn't used right now!

### CODE REVIEW

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java [hunk 1]
```diff
@@ -58,47 +47,19 @@
...
+    private final ImmutableList<LimitedChannel> channels;
 
     BalancedNodeSelectionStrategyChannel(
             ImmutableList<LimitedChannel> channels,
             Random random,
             Ticker ticker,
             TaggedMetricRegistry taggedMetrics,
             String channelName,
-            RttSampling samplingEnabled) {
+            RttSampling _samplingEnabled) {
```
Comment (iamdanfox) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java]: I'm intending to re-write this _after_ we finish the internal-ski-product use case.

Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java]: Might as well remove this flag entirely for now, can always plumb it through again later.

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java [hunk 2]
```diff
@@ -109,283 +70,33 @@
...
-
     @VisibleForTesting
     IntStream getScoresForTesting() {
-        return Arrays.stream(computeScores(rttSampler, channels)).mapToInt(SortableChannel::getScore);
+        return tracker.getScoresForTesting();
     }
 
     @Override
     public String toString() {
         return "BalancedNodeSelectionStrategyChannel{channels=" + channels + '}';
```
Comment (carterkozak) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java]: Let's add BalancedScoreTracker to the toString

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java [hunk 3]
```diff
@@ -58,47 +47,19 @@
...
             Random random,
             Ticker ticker,
             TaggedMetricRegistry taggedMetrics,
             String channelName,
-            RttSampling samplingEnabled) {
+            RttSampling _samplingEnabled) {
         Preconditions.checkState(channels.size() >= 2, "At least two channels required");
-        this.random = random;
-        this.clock = ticker;
-        this.rttSampler = samplingEnabled == RttSampling.DEFAULT_OFF ? null : new RttSampler(channels, clock);
```
Comment (carterkozak) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java:92]: I'd prefer to remove the RttSampler and related tests entirely, git won't forget about them.

**Comment by carterkozak** [comment]
lgtm aside from a few nits :+1:
I prefer to delete unused code entirely, after spending a few years in a large, old project, but if you plan to reimplement rtt very soon I could be swayed.

**Pull request merged**